### PR TITLE
Email: Send Poste activity metrics to New Relic

### PR DESCRIPTION
**Pull request by islemaster**
[PLC-336](https://codedotorg.atlassian.net/browse/PLC-336) Send the following metrics to New Relic every time we run the `deliver_poste_messages` task (every minute):

- `Custom/Poste/Sent`
  The number of emails sent on _this run_ of the cronjob.  Should never exceed 5000.
- `Custom/Poste/Abandoned`
  The total number of emails in `pegasus.poste_deliveries` with a `sent_at` value of `0`, indicating that we've failed to send the message and will not retry.  Currently hovering at ~50,000.
- `Custom/Poste/Queued`
  The total number of emails in `pegasus.poste_deliveries` with a null `sent_at` column, indicating we will attempt to send them on subsequent runs.  Should usually remain near-zero.

The `poste_deliveries` table holds about 30 million rows right now, but existing indexes make the queries used here very fast.

I'm mildly worried about rate limits for custom metrics on New Relic due to this task running every minute, but historically it's been very resilient about us throwing large amounts of data at it, and their limits seem to be mostly about number of unique metric/event keys sent, not how much data is sent per metric.

Follow-up work will be making a New Relic dashboard for monitoring Poste activity, and establishing what "normal" looks like, and possibly setting up alarms to catch future issues.

### CODE REVIEW

File: bin/cron/deliver_poste_messages [hunk 1]
```diff
@@ -342,14 +343,22 @@ def main
...
+  sent_count = POSTE_DB[:poste_deliveries].where(Sequel.lit('sent_at >= ?', started_at)).count
+
+  if CDO.newrelic_logging
+    # How many emails we sent on _this run_ of the cronjob
+    NewRelic::Agent.record_metric("Custom/Poste/Sent", sent_count)
+    # How many total abandoned emails we have
+    NewRelic::Agent.record_metric("Custom/Poste/Abandoned", abandon_count)
+    # How many emails are still queued for send on subsequent runs
+    NewRelic::Agent.record_metric("Custom/Poste/Queued", remaining_count)
+  end
```
Comment (breville) [inline bin/cron/deliver_poste_messages:357]: total nit: could order these the same as the variables are set, above

File: bin/cron/deliver_poste_messages [hunk 2]
```diff
@@ -342,14 +343,22 @@ def main
...
 
-  elapsed_seconds = ((Time.now - started_at) * 100.0).to_i / 100.0
-
-  unless queue_count == 0 && remaining_count == 0
+  unless queue_count == 0 && remaining_count == 0 && rack_env?(:development)
```
Comment (breville) [inline bin/cron/deliver_poste_messages]: just curious: what is the reason the code has always suppressed the message only when the counts are zero and we're on the development environment?

Comment (breville) [reply bin/cron/deliver_poste_messages]: oh actually, didn't we previously only `puts` on development?

Comment (hacodeorg) [reply bin/cron/deliver_poste_messages]: This is equivalent to `if queue_count != 0 || remaining_count != 0 || rack_env?(:development)` then `puts` isn't it?
(For some reasons every timeI see `unless` at the beginning of a block of code I have to translate it into an `if` block.)

**Comment by hacodeorg** [comment]
So what we're measuring are # of sent/abandon/remaining emails every minute? Will we have mostly quite time then several burst of activities? Can NewRelic aggregate data per hour or day or week to make it easier to detect pattern?

File: bin/cron/deliver_poste_messages (hunk 2 above)
Comment (islemaster) [reply bin/cron/deliver_poste_messages]: :man_facepalming: I definitely mixed up the `if` and `unless` cases here.  I'd actually love to just tear out this development-only code anyway - thoughts on that?

**Comment by islemaster** [comment]
Yes! We're setting these up as New Relic [metrics, not events](https://docs.newrelic.com/docs/using-new-relic/metrics/analyze-your-metrics/data-collection-metric-timeslice-event-data), so it's got a lot of power to aggregate these to whatever time-slice we choose and detect patterns.

A great recent example of this is the S3 LIST call instrumentation we did prior to Hour of Code this year.  [Here's the New Relic dashboard for that.](https://insights.newrelic.com/accounts/501463/dashboards/753432?duration=604800000)

Here's the lines where we actually record this metric:

https://github.com/code-dot-org/code-dot-org/blob/febafd86182344307ca99d909bc6da312423f7d4/shared/middleware/helpers/bucket_helper.rb#L515-L518

You can hunt for all the uses of `track_list_operation` to see how widespread this is. Notice we're passing a metric value of `1` for every call here. This data is used to generate this graph (and others), recording more than 65k events per hour with a one-hour timeslice in this view.

image

With the email data, I expect to measure "total value" for sent emails to get a sent-per-hour view or similar rollup, and a "max value" for abandoned / remaining since they are not cumulative values but instead we want to know the worst case in any given timeslice.

image

I'd like to see patterns for send rates, but also see our queue stay near-zero, and I want to see our abandon rate stop growing and maybe reduce it to zero over time.

File: bin/cron/deliver_poste_messages (hunk 2 above)
Comment (breville) [reply bin/cron/deliver_poste_messages]: The message seems useful for manual runs of this cron job, but not for any other case.  I wouldn't mind it being removed.

**Comment by hacodeorg** [comment]
Awesome, thanks for example Brad!

**Review by hacodeorg — APPROVED** [review]

**Comment by islemaster** [comment]
:cry: [I broke it.](https://app.honeybadger.io/projects/45435/faults/51272499)

> uninitialized constant NewRelic (NameError)

This shouldn't prevent any emails from sending since it's at the end of the script; it's just generating a lot of noise in Honeybadger.  I'm [squelching the errors by disabling this logging](https://github.com/code-dot-org/code-dot-org/commit/3f24a18efe8fc0e4c0356cc5bdd65444e66f1670) and will open a PR to re-enable it tomorrow.

**Pull request merged**